python/uds.py

```diff
@@ -376,39 +378,46 @@ def send(self, dat: bytes) -> None:
     self.rx_done = False
 
     if self.debug:
-      print(f"ISO-TP: REQUEST - 0x{bytes.hex(self.tx_dat)}")
+      print(f"ISO-TP: REQUEST - {hex(self._can_client.tx_addr)} 0x{bytes.hex(self.tx_dat)}")
     self._tx_first_frame()
 
   def _tx_first_frame(self) -> None:
     if self.tx_len < self.max_len:
       # single frame (send all bytes)
       if self.debug:
-        print("ISO-TP: TX - single frame")
+        print(f"ISO-TP: TX - {hex(self._can_client.tx_addr)} single frame")
       msg = (bytes([self.tx_len]) + self.tx_dat).ljust(self.max_len, b"\x00")
       self.tx_done = True
     else:
       # first frame (send first 6 bytes)
       if self.debug:
-        print("ISO-TP: TX - first frame")
+        print(f"ISO-TP: TX - {hex(self._can_client.tx_addr)} first frame")
       msg = (struct.pack("!H", 0x1000 | self.tx_len) + self.tx_dat[:self.max_len - 2]).ljust(self.max_len - 2, b"\x00")
     self._can_client.send([msg])
 
   def recv(self) -> Optional[bytes]:
     start_time = time.time()
+    self.rx_idx_prev = 0
     try:
       while True:
         for msg in self._can_client.recv():
           self._isotp_rx_next(msg)
           if self.tx_done and self.rx_done:
             return self.rx_dat
+          if self.rx_idx > self.rx_idx_prev:
+            start_time = time.time()
+            self.rx_idx_prev = self.rx_idx
         # no timeout indicates non-blocking
         if self.timeout == 0:
           return None
-        if time.time() - start_time > self.timeout:
+        if self.response_pending:
+          if time.time() - start_time > self.response_pending_timeout:
+            raise MessageTimeoutError("timeout waiting for pending response")
```

Comment:
it looks like `response_pending` is never reset to `False` but I think there might be a better way to build this:
* in `_uds_request` use a local variable to track if a response is pending
* add an optional param to this `recv()` method which specifies the timeout for only the first result from `_isotp_rx_next` (subsequent responses respect the normal timeout)

decreasing the scope of where variables are set/reset should decrease the chance of bugs in the future and make the code easier to understand

Reply:
Added `response_pending = False` after `_isotp_rx_next`